### PR TITLE
updating lucene updates to not include duplicate and skip indexEntry …

### PR DESCRIPTION
**Pull request by sophiabuell**
…logic

### CODE REVIEW

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java [hunk 1]
```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
...
+        if (newRecord != null) {
+            newRecordFields.addAll(getFields(root, newRecord, newRecord.getRecord()));
         }
+        if (oldRecord != null) {
+            oldRecordFields.addAll(getFields(root, oldRecord, oldRecord.getRecord()));
+        }
+
+        // evaluate which fields are changed
+        //Not sure if we want to allow deletion of fields. Or if the updated record will contain all fields not
+        // just the updated ones. Assuming here that the record will be complete if being updated as previously.
```
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:237]: Both records are complete (or `null`), yes.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java [hunk 2]
```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
...
+        List<DocumentEntry> updatedFields = newRecordFields;
+        updatedFields.removeAll(oldRecordFields);
```
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:239]: This assumes that `DocumentEntry` has a meaningful (non-reference) `equals`, which it does not right now.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java [hunk 3]
```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
...
-                    break;
-                case LONG:
-                    document.add(new LongPoint(fieldName, (long) value));
-                    break;
-                default:
-                    throw new RecordCoreArgumentException("Invalid type for lucene index field", "type", expression.getType());
+            // if old record has fields in the index and there are changes then delete old record
+            if (!oldRecordFields.isEmpty() && (!updatedFields.isEmpty() || newRecord == null)) {
+                Query query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(oldRecord.getPrimaryKey().pack()));
+                writer.deleteDocuments(query);
```
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:250]: Is this necessary when we call `addDocument` below? That is, without it, would there be two documents with the same fields or would the one below overwrite?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java [hunk 4]
```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
...
             }
+
+            // if there are changes then update the index with the new document.
+            if (!updatedFields.isEmpty()) {
+                byte[] primaryKey = newRecord.getPrimaryKey().pack();
+                BytesRef ref = new BytesRef(primaryKey);
+                document.add(new StoredField(PRIMARY_KEY_FIELD_NAME, ref));
+                document.add(new SortedDocValuesField(PRIMARY_KEY_SEARCH_NAME, ref));
+                for (DocumentEntry entry : newRecordFields) {
```
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:259]: `newRecordFields` and `updatedFields` are the same, so this is only the changed fields. Don't we need them all?

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java (hunk 3 above)
Comment (sophiabuell) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:250]: Without it there would be two documents with the same fields. Same as how there can be multiple fields with the exact same key/value.

Comment (sophiabuell) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:250]: it does not delete the previous one. It passes null to the delNode value in the update method.

Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:266]: ✅

**Review by MMcM — APPROVED** [review]

**Pull request merged**